twilight-gateway/src/shard.rs

```diff
@@ -0,0 +1,993 @@
+        let session_sequence = self.session.as_ref().map(Session::sequence);
+
+        if let Some(sequence) = override_sequence.or(session_sequence) {
+            let command = Heartbeat::new(sequence);
+            self.command(&command).await?;
+
+            self.latency.track_sent();
+        }
+
+        Ok(())
+    }
+
+    /// Identify a new session with the Discord gateway.
+    ///
+    /// # Errors
+    ///
+    /// Refer to [`command`][`Self::command`] for possible errors.
+    async fn identify(&mut self) -> Result<(), SendError> {
+        let properties = self
+            .config()
+            .identify_properties()
+            .cloned()
+            .unwrap_or_else(default_identify_properties);
+
+        let identify = Identify::new(IdentifyInfo {
+            compress: false,
+            large_threshold: self.config.large_threshold(),
+            intents: self.config.intents(),
+            properties,
+            shard: Some([self.id().number(), self.id().total()]),
+            presence: self.config.presence().cloned(),
+            token: self.config.token().to_owned(),
+        });
+
+        self.command(&identify).await
+    }
+
+    /// Process the buffer of the current websocket message to update the shard's
+    /// state.
+    ///
+    /// # Errors
+    ///
+    /// Returns a [`ProcessErrorType::Compression`] error type if the buffer
+    /// could not be read. This may be because the message was invalid or not
+    /// all frames have been received.
+    ///
+    /// Returns a [`ProcessErrorType::Deserializing`] error type if the gateway
+    /// event isn't a recognized structure, which may be the case for new or
+    /// undocumented events.
+    ///
+    /// Returns a [`ProcessErrorType::ParsingPayload`] error type if the buffer
+    /// isn't a valid [`GatewayEvent`]. This may happen if the opcode isn't
+    /// present.
+    ///
+    /// Returns a [`ProcessErrorType::SendingMessage`] error type if a Websocket
+    /// message couldn't be sent over the connection, which may be the case if
+    /// the connection isn't connected.
+    ///
+    /// [`GatewayEvent`]: twilight_model::gateway::event::GatewayEvent
+    async fn process(&mut self) -> Result<(), ProcessError> {
+        let buffer = match self.compression.message_mut() {
+            Ok(Some(buffer)) => buffer,
+            Ok(None) => return Ok(()),
+            Err(source) => return Err(ProcessError::from_compression(source)),
+        };
+
+        // Instead of returning the event type, we return whether the event type
+        // is a Ready event, which is the only one we handle. This gets around
+        // having both an immutable and mutable lifetime to the buffer.
+        let (raw_opcode, maybe_sequence, is_ready) = {
+            let json = String::from_utf8_lossy(buffer);
+            let deserializer = GatewayEventDeserializer::from_json(&json).ok_or(ProcessError {
```

Comment:
If `buffer` is not valid UTF-8 `String::from_utf8_lossy` will produce a string that will *ALWAYS* fail to deserialize so it's best to use `str::from_utf8` and return early on errors.